src/main/java/org/wso2/carbon/connector/connection/EmailConnectionManager.java

```diff
@@ -0,0 +1,163 @@
+/*
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.connector.connection;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.connector.exception.EmailConnectionException;
+import org.wso2.carbon.connector.exception.EmailConnectionPoolException;
+import org.wso2.carbon.connector.pojo.ConnectionConfiguration;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.lang.String.format;
+
+/**
+ * Manages email connections and connection pools
+ */
+public class EmailConnectionManager {
+
+    private static final Log log = LogFactory.getLog(EmailConnectionManager.class);
+
+    // Stores the connections used for SMTP against the connection name since no connection pooling is required for them
+    private Map<String, EmailConnection> connectionMap;
+    // Stores the connection pool created for protocols IMAP and POP3 against the connection name
+    private Map<String, EmailConnectionPool> connectionPoolMap;
+
+    private static EmailConnectionManager manager;
+
+    private EmailConnectionManager(){
+        this.connectionMap = new ConcurrentHashMap<>(new HashMap<>());
+        this.connectionPoolMap = new ConcurrentHashMap<>(new HashMap<>());
+    }
+
+    /**
+     * Gets Email Connection Manager
+     *
+     * @return EmailConnectionManager instance
+     */
+    public static synchronized EmailConnectionManager getEmailConnectionManager(){
```

Comment:
We better think again on making this method  synchronized. Will this make concurrent connection requests to loose overall concurrent processing?

Reply:
Using the synchronized block is identified to have several issues as per https://rules.sonarsource.com/java/tag/multi-threading/RSPEC-2168 and this is the suggested as the compliant solution as the newer JVMs is said to have improved the performance of `synchronized`.